accepted/2024/tensors/README.md

```diff
@@ -0,0 +1,424 @@
+
+### Base Interface
+
+At it's core there is an `ITensor<TSelf, T>` interface that forms the basis of interchange between types. This interface defines the minimal surface area expected for users to be able interact with a tensor.
+
+Different libraries support different sets of `T`. For example, some only support numeric `T`, some also support `string` or `bool`, and some even support more esoteric concepts like `DateTime`. .NET does not, today, have a system where we can say that members are available when a `T` meets a given constraint and so, much like with `generic math`, the number of interfaces we expose here "could" become quite expansive (potentially mirroring the generic math interface hierarchy). Extension methods do largely work, but they aren't possible today for `properties` or `operators` and so this somewhat limits our choices. However, `extension everything` has been a long standing request and this might be the right time to push for the support as it greatly simplifies our surface area here.
+
+To keep things simple, and in hopes of some support for extension `properties` and `operators`, this initially just defines the core tensor type which allows access to the data following a particular shape, to take slices of that data, and to reshape that data.
+
+```csharp
+namespace System.Numerics.Tensors;
+
+public interface ITensor<TSelf, T>
+    : IEnumerable<T>,
+      IEquatable<TSelf>,
+      IEqualityOperators<TSelf, TSelf, bool>
+    where TSelf : ITensor<TSelf, T>
+    where T : IEquatable<T>, IEqualityOperators<T, T, bool>
+{
+    // TODO: Determine if we can implement `IEqualityOperators<TSelf, T, bool>`.
+    // It looks like C#/.NET currently hits limitations here as it believes TSelf and T could be the same type
+    // Ideally we could annotate it such that they cannot be the same type and no conflicts would exist
+
+    static TSelf Empty { get; }
+
+    bool IsEmpty { get; }
+    bool IsPinned { get; }
+    int Rank { get; }
+    ReadOnlySpan<nint> Strides { get; }
+
+    ref T this[params ReadOnlySpan<nint> indices] { get; }
+
+    static implicit operator TensorSpan<T>(TSelf value);
+    static implicit operator TensorReadOnlySpan<T>(TSelf value);
+
+    TensorSpan<T> AsSpan(params ReadOnlySpan<Range<nint>> ranges);
+    ReadOnlyTensorSpan<T> AsReadOnlySpan(params ReadOnlySpan<Range<nint>> ranges);
+    ref T GetPinnableReference();
+    TSelf Slice(params ReadOnlySpan<Range<nint>> ranges);
+
+    void Clear();
+    void CopyTo(TensorSpan<T> destination);
+    void Fill(T value);
+    bool TryCopyTo<T>(TensorSpan<T> destination);
+
+    // IEnumerable
+
+    IEnumerator<T> GetEnumerator();
+
+    // IEquatable
+
+    bool Equals(TSelf other);
+
+    // IEqualityOperators
+
+    static bool operator ==(TSelf left, TSelf right);
+    static bool operator ==(TSelf left, T right);
+
+    static bool operator !=(TSelf left, TSelf right);
+    static bool operator !=(TSelf left, T right);
```

Comment:
How is equality defined? Rank, strides, and all elements are equal?

Reply:
Yes. It would then fail for something like comparing a `4x1` vs a `1x4` as the first is four rows vs 4 columns. A user would need to reshape to compare them.

Reply:
> It would then fail for something like comparing a `4x1` vs a `1x4`

Fail as in throwing an exception? Would that carry over to the `IEquatable<T>` implementation or the `Equals(object)` method? Convention for such methods is to return `false` for comparands whose type or shape doesn't match.

Reply:
Fail would mean returning false, as is required by IEquatable and typical for equality implementations